# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/shell/js.cpp
# Commit: 5c5cf716aa0b
# Full Hash: 5c5cf716aa0b01aecd8702254427825c9de1b8fa
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2023-06-08 03:42:08
# Regressor Bug: 1835876
# File Overlap Count: 1
# Description:
#   Bug 1835876 part 2 - Disable code write protection in content processes. r=nbp
#   
#   We've worked with the security and performance teams to re-evaluate the W^X policy
#   we have in place to mark JIT code memory pages as either writable or executable
#   (but not both).
# ==============================================================================

diff -r 028f981600d7 -r 5c5cf716aa0b js/src/shell/js.cpp
--- a/js/src/shell/js.cpp	Wed Jun 07 16:34:51 2023 +0000
+++ b/js/src/shell/js.cpp	Wed Jun 07 16:34:51 2023 +0000
@@ -11437,6 +11437,11 @@
       !op.addStringOption('\0', "spectre-mitigations", "on/off",
                           "Whether Spectre mitigations are enabled (default: "
                           "off, on to enable)") ||
+      !op.addStringOption('\0', "write-protect-code", "on/off",
+                          "Whether the W^X policy is enforced to mark JIT code "
+                          "pages as either writable or executable but never "
+                          "both at the same time (default: on, off to "
+                          "disable)") ||
       !op.addStringOption('\0', "cache-ir-stubs", "on/off/call",
                           "Use CacheIR stubs (default: on, off to disable, "
                           "call to enable work-in-progress call ICs)") ||
@@ -12156,6 +12161,16 @@
     }
   }
 
+  if (const char* str = op.getStringOption("write-protect-code")) {
+    if (strcmp(str, "on") == 0) {
+      jit::JitOptions.maybeSetWriteProtectCode(true);
+    } else if (strcmp(str, "off") == 0) {
+      jit::JitOptions.maybeSetWriteProtectCode(false);
+    } else {
+      return OptionFailure("write-protect-code", str);
+    }
+  }
+
   if (const char* str = op.getStringOption("ion-scalar-replacement")) {
     if (strcmp(str, "on") == 0) {
       jit::JitOptions.disableScalarReplacement = false;